src/v/storage/types.h

```diff
@@ -340,11 +340,14 @@ struct compaction_config {
     explicit compaction_config(
       model::timestamp upper,
       std::optional<size_t> max_bytes_in_log,
+      std::optional<model::offset> max_compact_offset,
```

Comment:
we already use `model::offset::max()` as an unset value (in stm_manager) so it's better to stick with it and not to introduce `optional`. making a type optional adds another dimension to the value space which we should keep in mind when we use the type.

Reply:
This particular API, `compaction_config` uses `optional`, so I thought it was more consistent. I did use `offset::max()` for the underlying value, though. I feel like nullopt for the test cases is more ergonomic, but I can change it to `...max()`..

Reply:
> we already use model::offset::max()

most of the code base uses `model::offset{}` (aka `::min`) to indicate an unset state. unless using `::max` is convenient because of a calculation that is based on `min(a, b)` then it is probably worth sticking to default initialization?